### PR TITLE
Drivers: sensors: Add support for AMS AS621x series

### PR DESCRIPTION
**Pull request by jtbaumann**
Added support for the AMS AS621x series of temperature sensors as a
variant of the TI TMP108 temperature sensor.

Signed-off-by: Jared Baumann <jared.baumann8@t-mobile.com>

### CODE REVIEW

**Comment by catch-twenty-two** [comment]
@gmarull @MaureenHelm @mbolivar rather than submitting a new driver for this chip, we wanted to reuse the TMP108 code I had submitted already since 90% of the functionality of AMS AS621x is the same. (just different registers) We weren't sure what the best way to do this was, but thought this route made sense. Please let us know what you think.

File: drivers/sensor/tmp108/tmp108.c [hunk 1]
```diff
@@ -202,13 +199,18 @@ static int tmp108_attr_set(const struct device *dev,
...
 	if (chan != SENSOR_CHAN_AMBIENT_TEMP && chan != SENSOR_CHAN_ALL) {
 		return -ENOTSUP;
 	}
 
 	switch ((int) attr) {
 	case SENSOR_ATTR_HYSTERESIS:
+#if DT_PROP(DT_INST(0, ti_tmp108), variant_as621x)
+		LOG_WRN("AS621x Series lacks Hysterisis setttings");
+		break;
+#endif
```
Comment (MaureenHelm) [inline drivers/sensor/tmp108/tmp108.c:213]: This won't work if there are multiple instances of the driver. Add an element to `struct tmp108_config` so you can look up the value on a per-instance basis.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please rebase and remove the merge commit

File: dts/bindings/sensor/ti,tmp108.yaml [hunk 2]
```diff
@@ -16,3 +17,8 @@ properties:
       description: |
         Identifies the ALERT signal, which is active-low open drain when
         produced by the sensor.
+  
+    variant-as621x:
```
Comment (gmarull) [inline dts/bindings/sensor/ti,tmp108.yaml:21]: I think the right way to solve this case is to make the driver support two compatibles, factorizing common code. cc: @mbolivar-nordic

Comment (mbolivar-nordic) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: Agreed

Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: thanks, @gmarull and @mbolivar-nordic.  Sorry just getting back from a long break on this, so that's what we attempted to do.  1 register set is included if there is an AMS621 included in the device tree otherwise the register set from the TMP108 is included. Just wanted to make sure the approach is sound before continuing on with it.

File: drivers/sensor/tmp108/tmp108.c (hunk 1 above)
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:213]: Thanks @MaureenHelm will add this

Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c]: @MaureenHelm I couldn't really think of an easy way to do this, so I just ended up requiring the user to configure this in the Kconfig by indicating whether they are compiling for a board with the as621 or the tmp108. At this point it would be a lot of work to allow both chipsets to operate on the same board using the same driver, so I think this is a good compromise.

File: dts/bindings/sensor/ti,tmp108.yaml (hunk 2 above)
Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: @gmarull @mbolivar-nordic done, please take a look when you have a few minutes.

**Comment by catch-twenty-two** [comment]
> Please rebase and remove the merge commit

Done

File: dts/bindings/sensor/ti,tmp108.yaml (hunk 2 above)
Comment (mbolivar-nordic) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: Did you forget to push, @catch-twenty-two ? I see no changes.

Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: Hi @mbolivar-nordic , I did make a bunch of changes. The drivers share a common .h file and the few registers definitions that are different are included depending on which driver is chosen in the KConfig example (line 22):
```
#if CONFIG_AS621x
#include "as621x_defs.h"
#else
#include "tmp108_defs.h"
#endif
```

Comment (mbolivar-nordic) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: > Did you have something else in mind?

Yes. Take another look at @gmarull's advice (emphasis mine):

> I think the right way to solve this case is to **make the driver support two compatibles**, factorizing common code. 

I.e., you need two different devicetree bindings, one for each compatible. The driver should support **both** compatibles, and importantly both **at the same time**.

**Comment by JordanYates** [comment]
How similar are these two sensors, beyond the fact that they both have a single config register and a result register?
* The sample rates aren't all the same
* The output data format is not the same
* AS621 has no hysteresis settings
* TMP108 has no consecutive fault settings

You mention `Enable driver for the TMP108 temperature sensor and it's variant the AMS AS621`, but the changes seem larger than a variant to me (I would expect accuracy changes across a variant, not different functionality).

File: drivers/sensor/tmp108/tmp108.h [hunk 3]
```diff
@@ -13,33 +14,30 @@
 #include <zephyr/drivers/gpio.h>
 #include <zephyr/drivers/sensor/tmp108.h>
 
-#define TI_TMP108_REG_TEMP		0x00   /** Temperature register */
-#define TI_TMP108_REG_CONF		0x01   /** Configuration register */
+#define TI_TMP108_REG_TEMP			0x00   /** Temperature register */
+#define TI_TMP108_REG_CONF			0x01   /** Configuration register */
```
Comment (mmahadevan108) [inline drivers/sensor/tmp108/tmp108.h:18]: This change is not needed

File: drivers/sensor/tmp108/tmp108.h [hunk 4]
```diff
@@ -13,33 +14,30 @@
...
+#endif
 
-#define TI_TMP108_CONF_M0	0x0100 /** Mode 1 configuration bit */
-#define TI_TMP108_CONF_M1	0x0200 /** Mode 2 configuration bit */
-#define TI_TMP108_CONF_CR0	0x2000 /** Conversion rate 1 configuration bit */
-#define TI_TMP108_CONF_CR1	0x4000 /** Conversion rate 2 configuration bit */
-#define TI_TMP108_CONF_POL	0x0080 /** Alert pin Polarity configuration bit */
-#define TI_TMP108_CONF_TM	0x0400 /** Thermostat mode setting bit */
-#define TI_TMP108_CONF_HYS1	0x0020 /** Temperature hysteresis config 1 bit  */
-#define TI_TMP108_CONF_HYS0	0x0010 /** Temperature hysteresis config 2 bit */
```
Comment (mmahadevan108) [inline drivers/sensor/tmp108/tmp108.h:29]: Does this need to be deleted?

**Comment by henrikbrixandersen** [comment]
I'd go for two separate drivers given the details presented so far.

**Comment by mbolivar-nordic** [comment]
I haven't looked at the details but either alternative (1 driver that supports 2 compatibles or 2 drivers) are fine with me depending on whatever makes the most sense to the sensor folks.

File: dts/bindings/sensor/ti,tmp108.yaml (hunk 2 above)
Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: @mbolivar-nordic Thanks for your help, I have reworked everything so this driver now supports multiple instances of both the tmp108 and as6212 on the same board, if my approach looks ok Ill go ahead and clean everything up.

Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: @gmarull fyi

File: drivers/sensor/tmp108/Kconfig [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
 menuconfig TMP108
 	bool "TMP108 Temperature Sensor"
 	depends on I2C
 	help
-	  Enable driver for the TMP108 temperature sensor.
+	  Enable driver for the TMP108 temperature sensor and/or it's variant
+	  the AMS AS621.
 
 if TMP108
-
```
Comment (mbolivar-nordic) [inline drivers/sensor/tmp108/Kconfig:13]: Please don't make unnecessary whitespace changes like this. They clutter up the diff for reviewers and violate the "a commit does one thing" principle. In fact, they are the canonical example of doing two things :).

Slide 14: https://static.sched.com/hosted_files/zephyr2022/a3/HOWTO_Get_Your_Zephyr_Patches_Merged_ZDS_2022.pdf

File: dts/bindings/sensor/ti,tmp108.yaml (hunk 2 above)
Comment (mbolivar-nordic) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: >  I have reworked everything so this driver now supports multiple instances of both the tmp108 and as6212 on the same board, if my approach looks ok Ill go ahead and clean everything up.

Did you forget to push something? I still see the problematic `variant-as621x` property.

Regardless of whether you create two drivers or create one driver that supports two compatibles, you still need two devicetree bindings.

Remember: a compatible property says what type of hardware the node represents. Having compatible be "ti,tmp108", but then adding a property that says "just kidding, this is actually an as621x", is not the right thing to do. Two parts --> two compatibles.

Comment (catch-twenty-two) [reply dts/bindings/sensor/ti,tmp108.yaml:21]: Ah okay, thanks for the information, I didn't realize that, will make the change

File: drivers/sensor/tmp108/Kconfig (hunk 5 above)
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/Kconfig:13]: sorry about that, will do

File: drivers/sensor/tmp108/tmp108.c [hunk 6]
```diff
@@ -202,99 +192,106 @@ static int tmp108_attr_set(const struct device *dev,
...
+	int32_t uval;
 
 	if (chan != SENSOR_CHAN_AMBIENT_TEMP && chan != SENSOR_CHAN_ALL) {
 		return -ENOTSUP;
 	}
 
 	switch ((int) attr) {
 	case SENSOR_ATTR_HYSTERESIS:
+		if (TI_TMP108_HYSTER_0_C(dev) == TI_TMP108_CONF_NA) {
+			LOG_WRN("AS621x Series lacks Hysterisis setttings");
```
Comment (uLipe) [inline drivers/sensor/tmp108/tmp108.c:204]: If sensor attribute is not supported, ,it should return -ENOTSUP instead a plain warning

File: samples/sensor/tmp108/src/main.c [hunk 7]
```diff
@@ -142,8 +142,14 @@ void main(void)
 	tmp108 = DEVICE_DT_GET_ANY(ti_tmp108);
 
 	if (!tmp108) {
-		printf("error: no tmp108 device found\n");
-		return;
+		printf("warning: tmp108 device not found checking for compatible ams device\n");
+
+		tmp108 = DEVICE_DT_GET_ANY(ams_as6212);
```
Comment (uLipe) [inline samples/sensor/tmp108/src/main.c:147]: Since this is a sample I think is worth to be nit here, I suggest to rename tmp108 variable, it is strange to reuse it to gather another device with a different PN, even if they're compatible.

People who refer samples in general are searching a place to start, so keep code simpler to understand is always a good choice there.

something like `sensor_device` maybe? or declare two device descriptors, one to each sensor type.

**Review by uLipe — CHANGES_REQUESTED** [review]
Some minor nits, otherwise, nice work!

File: drivers/sensor/tmp108/tmp108.c (hunk 6 above)
Comment (catch-twenty-two) [reply drivers/sensor/tmp108/tmp108.c:204]: good catch, done

File: samples/sensor/tmp108/src/main.c (hunk 7 above)
Comment (catch-twenty-two) [reply samples/sensor/tmp108/src/main.c]: done

**Comment by catch-twenty-two** [comment]
@MaureenHelm and @mbolivar-nordic clean up and requested changes have been implemented, please review when you have some time. Thank you

**Comment by catch-twenty-two** [comment]
> @MaureenHelm and @mbolivar-nordic clean up and requested changes have been implemented, please review when you have some time. Thank you

@MaureenHelm @gmarull @mbolivar-nordic ping

**Review by cfriedt — DISMISSED** [review]
LGTM

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by MaureenHelm — COMMENTED** [review]
Please address the CI failures. We had a lot of CI issues last week in main, so you might just need to rebase.

It would also be nice to split the sample changes to a separate commit.

Otherwise, the code changes LGTM

**Review by DhruvaG2000 — CHANGES_REQUESTED** [review]
Code LGTM,
Please rebase and check if the CI errors go away.

**Comment by stephanosio** [comment]
> Please rebase and check if the CI errors go away.

No, the CI failures here are due to an invalid change in this PR.

**Review by MaureenHelm — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@uLipe please revisit

**Review by DhruvaG2000 — APPROVED** [review]

**Review by uLipe — APPROVED** [review]
LGTM
